### PR TITLE
[pytorch][PR] Record FutureNCCL callback stream on CUDA caching allocator

### PR DESCRIPTION
**Pull request by wayi1**
Summary:
When calling `then()` from WorkNCCL, record the input data pointers in futureNCCLCallbackStream_ before the execution of the input callback.

Note that the recording cannot be directly added to the lambda used by addCallback in ProcessGroupNCCL.hpp. This is because the type of future value in that context is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Finally, a `record_stream_cb_` member is added to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `caffe2/torch/csrc/jit/python/pybind_utils.h`, where Python dependency is allowed. 

Currently record_stream_cb_ is not really customizable, and we assume the input ivalue of this callback is or can be casted into a tensor that has exactly one tensor. This callback will only be used by NCCL backend.

Test Plan
buck test @mode/dev-nosan caffe2/test/distributed:c10d -- ProcessGroupNCCLTest
buck test @mode/dev-nosan caffe2/test/distributed:c10d  -- test_accumulate_gradients_no_sync_allreduce_with_then_hook
buck test @mode/dev-nosan caffe2/test/distributed:c10d  -- test_ddp_comm_hook_allreduce_with_then_hook_nccl

Differential Revision: D23910257

### CODE REVIEW

**Comment by pritamdamania87** [comment]
@ngimel Could you help review this PR? We're following the convention from here: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L968

File: torch/lib/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -916,6 +918,13 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
       outputs_->size() == 1,
       "WorkNCCL's getFuture API is only supported for single-process single-device mode.");
   auto deviceIndex = (*outputs_)[0].device().index();
+  // Do not free the resluting tensors before their usages on
+  // futureNCCLCallbackStream_ finish.
+  for (const auto& tensor : *outputs_) {
+    c10::cuda::CUDACachingAllocator::recordStream(
+        tensor.storage().data_ptr(), *futureNCCLCallbackStreams_[deviceIndex]);
+  }
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.cpp:926]: Don't think we need recordStream here since we aren't actually running on the futureNCCLCallbackStream here.

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 2]
```diff
@@ -295,6 +296,13 @@ class ProcessGroupNCCL : public ProcessGroup {
       addCallback(std::bind(
           [&](std::function<at::IValue(void)> cb) {
             try {
+              // Do not free the input tensors before their usages on
+              // futureNCCLCallbackStream_ finish.
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:300]: There is only one tensor in the vector, so let's just do something like this to be explicit:

```
auto tensors = value_.toTensorVector();
TORCH_INTERNAL_ASSERT(1, tensors.size());
c10::cuda::CUDACachingAllocator::recordStream(
                    tensors[0].storage().data_ptr(), *futureNCCLCallbackStream_);
```

Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:300]: Done. The size of tensors is indeed 1.

However, realized that the change here cannot pass test_accumulate_gradients_no_sync_allreduce_with_then_hook which invokes then -- the execution fails after this , with an error of "Timing out after 100 seconds and killing subprocesses".

I even tried to place this added code after "fut->markCompleted(at::IValue(cb()));" or even right before returning the future in this lambda, and tried to comment out recordStream call. Didn't work.

Comment (pritamdamania87) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:300]: Can you debug to see why the test is timing out? You can probably add some log statements to the unit test to see where it is getting stuck. Another option is to attach gdb to the child processes when the test is stuck.

Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:300]: I added LOG statements to debug before asking actually, but all the logs in this function could be printed out, so it's not timed out in this function. Will try gdb.

Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp]: It turns out that the type of future value here is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Per offline discussion, added a `record_stream_cb_` member to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `get_future` binding in init.cpp, where Python dependency is allowed. record_stream_cb_ can be customized by caller via `setRecordStreamCallback`, so here the future value does not have to be a tensor vector.

File: torch/csrc/distributed/c10d/init.cpp [hunk 3]
```diff
@@ -520,7 +528,10 @@ Example::
       )")
       .def(py::init<>());
 
-  shared_ptr_class_<::c10d::TCPStore>(module, "TCPStore", store,
+  shared_ptr_class_<::c10d::TCPStore>(
+      module,
+      "TCPStore",
+      store,
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:534]: nit: Usually try to avoid formatting changes since it makes the PR harder to review. You can always send a separate PR with the formatting changes.

File: torch/csrc/distributed/c10d/init.cpp [hunk 4]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
...
+            if (!fut->hasRecordStreamCallback()) {
+              auto record_stream_cb = [](const at::IValue& value,
+                                         at::cuda::CUDAStream cuda_stream) {
+                if (value.isTensorList() || value.isPyObject()) {
+                  pybind11::gil_scoped_acquire gil;
+                  std::vector<at::Tensor> tensors;
+                  if (value.isTensorList()) {
+                    tensors = std::move(value.toTensorVector());
+                  } else {
+                    py::object obj = torch::jit::toPyObject(value);
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:1028]: We probably need to acquire GIL only for this codepath? Can we move `pybind11::gil_scoped_acquire gil;` here?

File: torch/csrc/distributed/c10d/init.cpp [hunk 5]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
...
+                            .toTensorVector());
+                  }
+                  TORCH_INTERNAL_ASSERT(1, tensors.size());
+                  c10::cuda::CUDACachingAllocator::recordStream(
+                      tensors[0].storage().data_ptr(), cuda_stream);
+                }
+              };
+              fut->setRecordStreamCallback(record_stream_cb);
+            }
+            fut->recordStream();
```
Comment (pritamdamania87) [inline torch/csrc/distributed/c10d/init.cpp:1041]: recordStream should not be called here, it should be done only just before we call execute an OP on a new stream. Basically it should be done just before `streamGuard` in `addCallback()`

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 6]
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
       return !value_.isNone();
     }
 
+    std::shared_ptr<at::cuda::CUDAStream> getfutureNCCLCallbackStream() {
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:336]: This method is not used.

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 7]
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
...
+            record_stream_cb) {
+      record_stream_cb_ = std::move(record_stream_cb);
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
+      // Do not free the underlying data storage of value_ before its usage on
+      // futureNCCLCallbackStream_ finish.
+      record_stream_cb_(value_, *futureNCCLCallbackStream_);
+    }
```
Comment (pritamdamania87) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:355]: You probably don't need a separate method for this if you're doing it in `addCallback()` instead.

**Review by pritamdamania87 — CHANGES_REQUESTED** [review]
Requesting changes since I think `recordStream` should be done inside addCallback(). Also, can we run the DDP/HPC benchmarks with this change (with fp16 hook enabled) to ensure it doesn't cause any issues? You can probably verify QPS and peak memory to see they're not affected significantly.

File: torch/csrc/distributed/c10d/init.cpp (hunk 3 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp:534]: Yeah, realized that. Will be more careful next time.

File: torch/csrc/distributed/c10d/init.cpp [hunk 8]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
...
+            auto fut = dynamic_cast<::c10d::ProcessGroupNCCL::FutureNCCL*>(
+                work.getFuture().get());
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
+              auto record_stream_cb = [](const at::IValue& value,
+                                         at::cuda::CUDAStream cuda_stream) {
+                if (value.isTensorList() || value.isPyObject()) {
+                  pybind11::gil_scoped_acquire gil;
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:1023]: I think this function already holds GIL (because there is no gil_scoped_release guard in the definition)? You can probably check this by using `PyGILState_Check`. If so, I think re-acquiring the GIL here could result in some issues.

File: torch/csrc/distributed/c10d/init.cpp [hunk 9]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
...
+                  if (value.isTensorList()) {
+                    tensors = std::move(value.toTensorVector());
+                  } else {
+                    py::object obj = torch::jit::toPyObject(value);
+                    tensors = std::move(
+                        torch::jit::toIValue(
+                            obj, c10::ListType::create(c10::TensorType::get()))
+                            .toTensorVector());
+                  }
+                  TORCH_INTERNAL_ASSERT(1, tensors.size());
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:1034]: Is this the correct use of the API? Shouldn't it be `TORCH_INTERNAL_ASSERT(tensors.size() == 1, "expected tensors.size to be 1")` or something like that?

File: torch/csrc/distributed/c10d/init.cpp [hunk 10]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
           "get_future",
           [](::c10d::ProcessGroup::Work& work)
               -> std::shared_ptr<jit::PythonFutureWrapper> {
+#ifdef USE_C10D_NCCL
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:1013]: Might just be my confusion on this compiler guard. But isn't this true if we compile with NCCL enabled, but the user could still choose the Gloo backend at runtime? What sort of error message do we get if this codepath is executed with the Gloo backend? Could we add a test to ensure something reasonable is raised in this case (can be in a separate diff).

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 11]
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
...
+
+    void setRecordStreamCallback(
+        std::function<void(const at::IValue&, at::cuda::CUDAStream)>
+            record_stream_cb) {
+      record_stream_cb_ = std::move(record_stream_cb);
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
```
Comment (rohan-varma) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:351]: nit: I believe this should be `TORCH_INTERNAL_ASSERT` as it is a framework bug if this function is called with an invalid record_stream_cb_. 

I believe TORCH_CHECK is used for error validation coming from user input/code, and TORCH_INTERNAL_ASSERT for framework validation.

File: torch/csrc/distributed/c10d/init.cpp [hunk 12]
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
...
+            auto fut = dynamic_cast<::c10d::ProcessGroupNCCL::FutureNCCL*>(
+                work.getFuture().get());
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/init.cpp:1019]: Could there be race conditions where different threads could try to set this at the same time? Do we support user calling this code from multithreaded environment?

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 13]
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
...
+      // futureNCCLCallbackStream_ finish.
+      record_stream_cb_(value_, *futureNCCLCallbackStream_);
+    }
+
    private:
     at::IValue value_;
     c10::DeviceIndex deviceIndex_;
     std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents_;
     std::shared_ptr<at::cuda::CUDAStream> futureNCCLCallbackStream_;
+    std::function<void(const at::IValue&, at::cuda::CUDAStream)>
```
Comment (rohan-varma) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:362]: Nit: up to your preference, but you could do `using recordStreamCallback = std::function<void (....` to avoid having to write this twice. Although since it's used only twice not sure if it's worth it.

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 14]
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
...
+
+    void setRecordStreamCallback(
+        std::function<void(const at::IValue&, at::cuda::CUDAStream)>
+            record_stream_cb) {
+      record_stream_cb_ = std::move(record_stream_cb);
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
+      // Do not free the underlying data storage of value_ before its usage on
```
Comment (rohan-varma) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:352]: cc @pritamdamania87 , is there anyway we could write a test that would fail without this record stream callback logic?

Comment (pritamdamania87) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:352]: I think it might be possible but tricky to do. It probably has to be a C++ test and interact with CUDACachingAllocator in a way that the caching allocator frees up the data, but we're still using the data in futureNCCLCallbackStream.

File: torch/csrc/distributed/c10d/init.cpp (hunk 9 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp:1034]: Fixed.

File: torch/csrc/distributed/c10d/init.cpp (hunk 10 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp:1013]: There shouldn't be any side effect on other backend, as no code will be executed.

File: torch/lib/c10d/ProcessGroupNCCL.hpp (hunk 11 above)
Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:351]: Fixed.

File: torch/csrc/distributed/c10d/init.cpp (hunk 12 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp:1019]: The asynchronous threads/streams on the same device actually will be chained in the order of creation time (i.e., forked by the call of then). See the comment on addCallback function in ProcessGroupNCCL.hpp.

File: torch/csrc/distributed/c10d/init.cpp (hunk 8 above)
Comment (SciPioneer) [reply torch/csrc/distributed/c10d/init.cpp:1023]: The new design moved this part to pybind_utils.h

File: torch/lib/c10d/ProcessGroupNCCL.hpp (hunk 13 above)
Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:362]: I also had the same thought. After realizing that `std::function<at::IValue(void)> cb` has also appeared twice in this file, I decided not to create a separate alias.

File: torch/lib/c10d/ProcessGroupNCCL.hpp (hunk 14 above)
Comment (SciPioneer) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:352]: Currently I hacked the test file to locally verify that in a test case like fut.then(mult).then(div), CUDACachingAllocator::recordStream has been invoked by the first two futures, respectively.

Agree that there could be a better approach to create a C++ unit test that can allow CUDACachingAllocator to free up the cached data.

**Comment by SciPioneer** [comment]
> Requesting changes since I think `recordStream` should be done inside addCallback(). Also, can we run the DDP/HPC benchmarks with this change (with fp16 hook enabled) to ensure it doesn't cause any issues? You can probably verify QPS and peak memory to see they're not affected significantly.

Per offline discussion, ran the callback right before `addCallback` is invoked.

File: aten/src/ATen/core/ivalue_inl.h [hunk 15]
```diff
@@ -411,6 +411,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     return fut;
   }
 
+  // Since this file cannot import CUDA depedency, use void* to represent CUDA
+  // stream, and will cast it into the correct stream type whenever use it.
+  virtual void setRecordStreamCallback(
```
Comment (pritamdamania87) [inline aten/src/ATen/core/ivalue_inl.h:416]: You can change the signature of this method to the following to address the CUDA stream issue:

```
template< typename T>
virtual void setRecordStreamCallback(T record_stream_cb) {}
```

This is also slightly more efficient than std::function and I think you can pass in a lambda here for `record_stream_cb` which accepts a CUDA stream as input.

File: aten/src/ATen/core/ivalue_inl.h [hunk 16]
```diff
@@ -411,6 +411,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     return fut;
   }
 
+  // Since this file cannot import CUDA depedency, use void* to represent CUDA
+  // stream, and will cast it into the correct stream type whenever use it.
```
Comment (pritamdamania87) [inline aten/src/ATen/core/ivalue_inl.h:415]: Add a comment here mentioning the default implementation is a no-op since it doesn't deal with any CUDA streams.

File: aten/src/ATen/core/ivalue_inl.h (hunk 15 above)
Comment (SciPioneer) [reply aten/src/ATen/core/ivalue_inl.h:416]: Please note that setRecordStreamCallback is defined inside ivalue::Future. Do we want to make ivalue::Future a template class? If so, how shall we instantiate T for other backends like Gloo and MPI?

Another concern: I am a fan of template class, but making this core class as a template class for only such a minor method may not justify the loss of debug-ability to some extent, due to the notorious error messages. As I can see, this class seems to be updated quite a few times recently.

Comment (SciPioneer) [reply aten/src/ATen/core/ivalue_inl.h:418]: Alternatively, what about replacing void* and reinterpret_cast with std::any and any_cast? This is a C++17 feature, and I am glad that our codebase can support C++17.

Comment (SciPioneer) [reply aten/src/ATen/core/ivalue_inl.h:418]: Per offline discussion, used c10::Stream to construct CUDAStream on the fly and avoid the usage of reinterpret_cast.

**Review by pritamdamania87 — COMMENTED** [review]
Changes LGTM, although there are a lot of CI failures. Can you fix those?

**Review by pritamdamania87 — APPROVED** [review]
